File: src/db/queries/trips.ts
import { getDatabase } from '../database.js';

export interface Trip {
  id: string;
  load_owner_id: string;
  load_owner_name: string;
  load_owner_logo?: string;
  load_owner_rating?: number;
  client_company?: string;
  client_logo?: string;
  transporter_id?: string;
  transporter_name?: string;
  origin: string;
  destination: string;
  distance: number;
  load_type: string;
  weight: number;
  amount: number;
  interest_rate?: number;
  maturity_days?: number;
  risk_level?: 'low' | 'medium' | 'high';
  insurance_status: boolean;
  status: 'pending' | 'escrowed' | 'funded' | 'in_transit' | 'completed' | 'cancelled';
  lender_id?: string;
  lender_name?: string;
  created_at: string;
  funded_at?: string;
  completed_at?: string;
}

export interface TripBid {
  id: string;
  trip_id: string;
  lender_id: string;
  lender_name: string;
  amount: number;
  interest_rate: number;
  created_at: string;
}

export interface TripDocument {
  id: string;
  trip_id: string;
  document_type: 'bilty' | 'ewaybill' | 'advance_invoice' | 'pod' | 'final_invoice';
  document_data: string;
  uploaded_at: string;
  uploaded_by: string;
}

export interface CreateTripInput {
  load_owner_id: string;
  load_owner_name: string;
  load_owner_logo?: string;
  load_owner_rating?: number;
  client_company?: string;
  client_logo?: string;
  origin: string;
  destination: string;
  distance: number;
  load_type: string;
  weight: number;
  amount: number;
  interest_rate?: number;
  maturity_days?: number;
  risk_level?: 'low' | 'medium' | 'high';
  insurance_status?: boolean;
}

/**
 * Get trip by ID with bids and documents
 */
export const getTrip = async (id: string): Promise<(Trip & { bids?: TripBid[], documents?: Record<string, string> }) | null> => {
  const db = await getDatabase();
  const result = await db.query('SELECT * FROM trips WHERE id = $1', [id]);
  const trip = result.rows[0];

  if (!trip) return null;

  // Get bids
  const bidsResult = await db.query('SELECT * FROM trip_bids WHERE trip_id = $1 ORDER BY created_at DESC', [id]);
  const bids = bidsResult.rows as TripBid[];

  // Get documents
  const docsResult = await db.query('SELECT * FROM trip_documents WHERE trip_id = $1', [id]);
  const docs = docsResult.rows as TripDocument[];
  const documents: Record<string, string> = {};
  docs.forEach(doc => {
    documents[doc.document_type] = doc.document_data;
  });

  // Convert numeric fields to numbers
  return {
    ...trip,
    distance: Number(trip.distance),
    weight: Number(trip.weight),
    amount: Number(trip.amount),
    interest_rate: trip.interest_rate ? Number(trip.interest_rate) : undefined,
    maturity_days: trip.maturity_days ? Number(trip.maturity_days) : undefined,
    load_owner_rating: trip.load_owner_rating ? Number(trip.load_owner_rating) : undefined,
    bids: bids.length > 0 ? bids.map(bid => ({
      ...bid,
      amount: Number(bid.amount),
      interest_rate: Number(bid.interest_rate),
    })) : undefined,
    documents: Object.keys(documents).length > 0 ? documents : undefined,
  };
};

/**
 * Get all trips with bids and documents
 */
export const getAllTrips = async (): Promise<(Trip & { bids?: TripBid[], documents?: Record<string, string> })[]> => {
  const db = await getDatabase();
  const result = await db.query('SELECT * FROM trips ORDER BY created_at DESC');
  const trips = result.rows as Trip[];

  try {
    // Get all bids for all trips in a single query for efficiency
    const bidsResult = await db.query('SELECT * FROM trip_bids ORDER BY created_at DESC');
    const allBids = bidsResult.rows as TripBid[];

    // Get all documents for all trips in a single query for efficiency
    const docsResult = await db.query('SELECT * FROM trip_documents');
    const allDocs = docsResult.rows as TripDocument[];

    // Group bids by trip_id and convert numeric fields
    const bidsByTrip = allBids.reduce((acc, bid) => {
      if (!acc[bid.trip_id]) {
        acc[bid.trip_id] = [];
      }
      acc[bid.trip_id].push({
        ...bid,
        amount: Number(bid.amount),
        interest_rate: Number(bid.interest_rate),
      });
      return acc;
    }, {} as Record<string, TripBid[]>);

    // Group documents by trip_id
    const docsByTrip = allDocs.reduce((acc, doc) => {
      if (!acc[doc.trip_id]) {
        acc[doc.trip_id] = {};
      }
      acc[doc.trip_id][doc.document_type] = doc.document_data;
      return acc;
    }, {} as Record<string, Record<string, string>>);

    // Add bids and documents to each trip and convert numeric fields
    return trips.map(trip => ({
      ...trip,
      distance: Number(trip.distance),
      weight: Number(trip.weight),
      amount: Number(trip.amount),
      interest_rate: trip.interest_rate ? Number(trip.interest_rate) : undefined,
      maturity_days: trip.maturity_days ? Number(trip.maturity_days) : undefined,
      load_owner_rating: trip.load_owner_rating ? Number(trip.load_owner_rating) : undefined,
      bids: bidsByTrip[trip.id] || undefined,
      documents: docsByTrip[trip.id] && Object.keys(docsByTrip[trip.id]).length > 0 ? docsByTrip[trip.id] : undefined,
    }));
  } catch (error) {
    console.error('Error loading bids and documents:', error);
    // Return trips with numeric conversions even if bids/documents fail
    return trips.map(trip => ({
      ...trip,
      distance: Number(trip.distance),
      weight: Number(trip.weight),
      amount: Number(trip.amount),
      interest_rate: trip.interest_rate ? Number(trip.interest_rate) : undefined,
      maturity_days: trip.maturity_days ? Number(trip.maturity_days) : undefined,
      load_owner_rating: trip.load_owner_rating ? Number(trip.load_owner_rating) : undefined,
    }));
  }
};

/**
 * Get trips by status with documents
 */
export const getTripsByStatus = async (status: Trip['status']): Promise<(Trip & { documents?: Record<string, string> })[]> => {
  const db = await getDatabase();
  const result = await db.query('SELECT * FROM trips WHERE status = $1 ORDER BY created_at DESC', [status]);
  const trips = result.rows as Trip[];

  // Get all documents for these trips
  const tripIds = trips.map(t => t.id);
  if (tripIds.length === 0) return [];

  const docsResult = await db.query('SELECT * FROM trip_documents WHERE trip_id = ANY($1)', [tripIds]);
  const allDocs = docsResult.rows as TripDocument[];

  // Group documents by trip_id
  const docsByTrip = allDocs.reduce((acc, doc) => {
    if (!acc[doc.trip_id]) {
      acc[doc.trip_id] = {};
    }
    acc[doc.trip_id][doc.document_type] = doc.document_data;
    return acc;
  }, {} as Record<string, Record<string, string>>);

  return trips.map(trip => ({
    ...trip,
    distance: Number(trip.distance),
    weight: Number(trip.weight),
    amount: Number(trip.amount),
    interest_rate: trip.interest_rate ? Number(trip.interest_rate) : undefined,
    maturity_days: trip.maturity_days ? Number(trip.maturity_days) : undefined,
    load_owner_rating: trip.load_owner_rating ? Number(trip.load_owner_rating) : undefined,
    documents: docsByTrip[trip.id] && Object.keys(docsByTrip[trip.id]).length > 0 ? docsByTrip[trip.id] : undefined,
  }));
};

/**
 * Get trips by load owner with documents
 */
export const getTripsByLoadOwner = async (loadOwnerId: string): Promise<(Trip & { documents?: Record<string, string> })[]> => {
  const db = await getDatabase();
  const result = await db.query('SELECT * FROM trips WHERE load_owner_id = $1 ORDER BY created_at DESC', [loadOwnerId]);
  const trips = result.rows as Trip[];

  // Get all documents for these trips
  const tripIds = trips.map(t => t.id);
  if (tripIds.length === 0) return [];

  const docsResult = await db.query('SELECT * FROM trip_documents WHERE trip_id = ANY($1)', [tripIds]);
  const allDocs = docsResult.rows as TripDocument[];

  // Group documents by trip_id
  const docsByTrip = allDocs.reduce((acc, doc) => {
    if (!acc[doc.trip_id]) {
      acc[doc.trip_id] = {};
    }
    acc[doc.trip_id][doc.document_type] = doc.document_data;
    return acc;
  }, {} as Record<string, Record<string, string>>);

  return trips.map(trip => ({
    ...trip,
    distance: Number(trip.distance),
    weight: Number(trip.weight),
    amount: Number(trip.amount),
    interest_rate: trip.interest_rate ? Number(trip.interest_rate) : undefined,
    maturity_days: trip.maturity_days ? Number(trip.maturity_days) : undefined,
    load_owner_rating: trip.load_owner_rating ? Number(trip.load_owner_rating) : undefined,
    documents: docsByTrip[trip.id] && Object.keys(docsByTrip[trip.id]).length > 0 ? docsByTrip[trip.id] : undefined,
  }));
};

/**
 * Get trips by lender with documents
 */
export const getTripsByLender = async (lenderId: string): Promise<(Trip & { documents?: Record<string, string> })[]> => {
  const db = await getDatabase();
  const result = await db.query('SELECT * FROM trips WHERE lender_id = $1 ORDER BY created_at DESC', [lenderId]);
  const trips = result.rows as Trip[];

  // Get all documents for these trips
  const tripIds = trips.map(t => t.id);
  if (tripIds.length === 0) return [];

  const docsResult = await db.query('SELECT * FROM trip_documents WHERE trip_id = ANY($1)', [tripIds]);
  const allDocs = docsResult.rows as TripDocument[];

  // Group documents by trip_id
  const docsByTrip = allDocs.reduce((acc, doc) => {
    if (!acc[doc.trip_id]) {
      acc[doc.trip_id] = {};
    }
    acc[doc.trip_id][doc.document_type] = doc.document_data;
    return acc;
  }, {} as Record<string, Record<string, string>>);

  return trips.map(trip => ({
    ...trip,
    distance: Number(trip.distance),
    weight: Number(trip.weight),
    amount: Number(trip.amount),
    interest_rate: trip.interest_rate ? Number(trip.interest_rate) : undefined,
    maturity_days: trip.maturity_days ? Number(trip.maturity_days) : undefined,
    load_owner_rating: trip.load_owner_rating ? Number(trip.load_owner_rating) : undefined,
    documents: docsByTrip[trip.id] && Object.keys(docsByTrip[trip.id]).length > 0 ? docsByTrip[trip.id] : undefined,
  }));
};

/**
 * Create trip
 */
export const createTrip = async (input: CreateTripInput): Promise<Trip> => {
  const db = await getDatabase();
  const id = `trip-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;

  await db.query(`
    INSERT INTO trips (
      id, load_owner_id, load_owner_name, load_owner_logo, load_owner_rating,
      client_company, client_logo, origin, destination, distance, load_type,
      weight, amount, interest_rate, maturity_days, risk_level, insurance_status, status
    ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, 'pending')
  `, [
    id,
    input.load_owner_id,
    input.load_owner_name,
    input.load_owner_logo || null,
    input.load_owner_rating || null,
    input.client_company || null,
    input.client_logo || null,
    input.origin,
    input.destination,
    input.distance,
    input.load_type,
    input.weight,
    input.amount,
    input.interest_rate || null,
    input.maturity_days || 30,
    input.risk_level || 'low',
    input.insurance_status || false
  ]);

  const trip = await getTrip(id);
  if (!trip) {
    throw new Error('Failed to create trip');
  }

  return trip;
};

/**
 * Update trip
 */
export const updateTrip = async (id: string, updates: Partial<Trip & { documents?: Record<string, string> }>): Promise<Trip | null> => {
  const db = await getDatabase();

  const trip = await getTrip(id);
  if (!trip) return null;

  // Extract and handle documents separately
  const { documents, ...tripUpdates } = updates as any;

  // Handle document updates if provided
  if (documents && typeof documents === 'object') {
    const validDocTypes: Array<'bilty' | 'ewaybill' | 'advance_invoice' | 'pod' | 'final_invoice'> =
      ['bilty', 'ewaybill', 'advance_invoice', 'pod', 'final_invoice'];

    for (const docType of validDocTypes) {
      if (documents[docType]) {
        // Use 'system' as uploadedBy for now, or extract from context if available
        await uploadDocument(id, docType, documents[docType], trip.load_owner_id || 'system');
      }
    }
  }

  const fields: string[] = [];
  const values: any[] = [];
  let paramIndex = 1;

  Object.entries(tripUpdates).forEach(([key, value]) => {
    if (value !== undefined && key !== 'id' && key !== 'documents') {
      fields.push(`${key} = $${paramIndex}`);
      values.push(value);
      paramIndex++;
    }
  });

  if (fields.length > 0) {
    values.push(id);

    await db.query(`
      UPDATE trips SET ${fields.join(', ')} WHERE id = $${paramIndex}
    `, values);
  }

  return await getTrip(id);
};

/**
 * Add bid to trip
 */
export const addBid = async (tripId: string, lenderId: string, lenderName: string, amount: number, interestRate: number): Promise<TripBid> => {
  const db = await getDatabase();
  const id = `bid-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;

  await db.query(`
    INSERT INTO trip_bids (id, trip_id, lender_id, lender_name, amount, interest_rate)
    VALUES ($1, $2, $3, $4, $5, $6)
  `, [id, tripId, lenderId, lenderName, amount, interestRate]);

  const result = await db.query('SELECT * FROM trip_bids WHERE id = $1', [id]);
  return result.rows[0] as TripBid;
};

/**
 * Update bid with contract reference
 */
export const updateBidContract = async (bidId: string, contractId: string): Promise<void> => {
  const db = await getDatabase();
  await db.query(`
    UPDATE trip_bids
    SET contract_id = $1, has_contract = TRUE
    WHERE id = $2
  `, [contractId, bidId]);
};

/**
 * Upload document
 */
export const uploadDocument = async (
  tripId: string,
  documentType: 'bilty' | 'ewaybill' | 'advance_invoice' | 'pod' | 'final_invoice',
  documentData: string,
  uploadedBy: string
): Promise<TripDocument> => {
  const db = await getDatabase();
  const id = `doc-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;

  // Delete existing document of same type
  await db.query('DELETE FROM trip_documents WHERE trip_id = $1 AND document_type = $2', [tripId, documentType]);

  await db.query(`
    INSERT INTO trip_documents (id, trip_id, document_type, document_data, uploaded_by)
    VALUES ($1, $2, $3, $4, $5)
  `, [id, tripId, documentType, documentData, uploadedBy]);

  const result = await db.query('SELECT * FROM trip_documents WHERE id = $1', [id]);
  return result.rows[0] as TripDocument;
};

/**
 * Get bids for trip
 */
export const getTripBids = async (tripId: string): Promise<TripBid[]> => {
  const db = await getDatabase();
  const result = await db.query('SELECT * FROM trip_bids WHERE trip_id = $1 ORDER BY created_at DESC', [tripId]);
  return result.rows;
};

/**
 * Delete trip
 */
export const deleteTrip = async (id: string): Promise<boolean> => {
  const db = await getDatabase();
  const result = await db.query('DELETE FROM trips WHERE id = $1', [id]);
  return result.rowCount > 0;
};

/**
 * Bulk create trips - input format for bulk creation
 */
export interface BulkTripInput {
  ewayBillNumber: string;
  pickup: string;
  destination: string;
  sender?: string;
  receiver: string;
  transporter: string;
  loanAmount: number;
  loanInterestRate: number;
  maturityDays: number;
  distance: number;
  loadType: string;
  weight: number;
}

/**
 * Result of bulk trip creation
 */
export interface BulkCreateResult {
  success: boolean;
  created: number;
  failed: number;
  errors: Array<{
    index: number;
    data: BulkTripInput;
    error: string;
  }>;
}

/**
 * Create trips in bulk from external data format
 */
export const bulkCreateTrips = async (
  trips: BulkTripInput[],
  findLoadOwnerByName: (name: string) => Promise<any>,
  findTransporterByName: (name: string) => Promise<any>,
  getCompanyInfo: (key: string) => any
): Promise<BulkCreateResult> => {
  const db = await getDatabase();
  const result: BulkCreateResult = {
    success: true,
    created: 0,
    failed: 0,
    errors: []
  };

  for (let i = 0; i < trips.length; i++) {
    const tripData = trips[i];

    try {
      // Try to get company info from companyDatabase first
      let receiverCompanyInfo = null;
      let loadOwner = null;

      if (tripData.receiver) {
        // Try to get from companyInfo database
        receiverCompanyInfo = getCompanyInfo(tripData.receiver);

        // Try to find load owner by the full company name or the key
        if (receiverCompanyInfo) {
          loadOwner = await findLoadOwnerByName(receiverCompanyInfo.name);
        }

        // If not found, try with the provided name directly
        if (!loadOwner) {
          loadOwner = await findLoadOwnerByName(tripData.receiver);
        }

        // If still not found, just use receiver as client company (not an error)
        if (!loadOwner) {
          console.log(`Load owner "${tripData.receiver}" not found in users database, using as client company`);
        }
      }

      // Find transporter - try companyInfo database first, then users database
      let transporterCompanyInfo = getCompanyInfo(tripData.transporter);
      let transporter = null;

      // Try to find transporter by full company name from companyInfo
      if (transporterCompanyInfo) {
        transporter = await findTransporterByName(transporterCompanyInfo.name);
      }

      // If not found, try with the provided name directly
      if (!transporter) {
        transporter = await findTransporterByName(tripData.transporter);
      }

      // If still not found, throw error (transporter is required)
      if (!transporter) {
        throw new Error(`Transporter "${tripData.transporter}" not found in database. Please use "Rolling Radius" or add the transporter to users database.`);
      }

      // Optional: sender is the client company (consignor)
      let clientCompany = tripData.sender || (tripData.receiver && !loadOwner ? tripData.receiver : null);
      let clientLogo = receiverCompanyInfo?.logo || null;

      // Create trip
      const id = `trip-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;

      //If no load owner found, we need a placeholder or use the transporter as load owner temporarily
      const effectiveLoadOwner = loadOwner || {
        id: transporter.id, // Use transporter as fallback
        name: tripData.receiver || 'Unknown',
        company: tripData.receiver,
        company_logo: receiverCompanyInfo?.logo || null,
        user_logo: null,
        rating: receiverCompanyInfo?.rating || null
      };

      await db.query(`
        INSERT INTO trips (
          id, load_owner_id, load_owner_name, load_owner_logo, load_owner_rating,
          client_company, client_logo, transporter_id, transporter_name,
          origin, destination, distance, load_type,
          weight, amount, interest_rate, maturity_days, risk_level, insurance_status, status
        ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, 'pending')
      `, [
        id,
        effectiveLoadOwner.id,
        effectiveLoadOwner.company || effectiveLoadOwner.name,
        effectiveLoadOwner.company_logo || effectiveLoadOwner.user_logo || null,
        effectiveLoadOwner.rating || null,
        clientCompany,
        clientLogo,
        transporter.id,
        transporter.company || transporter.name,
        tripData.pickup,
        tripData.destination,
        tripData.distance,
        tripData.loadType,
        tripData.weight,
        tripData.loanAmount,
        tripData.loanInterestRate,
        tripData.maturityDays,
        'medium', // Default risk level
        true // Default insurance status
      ]);

      result.created++;
    } catch (error: any) {
      result.failed++;
      result.errors.push({
        index: i,
        data: tripData,
        error: error.message || 'Unknown error occurred'
      });
      console.error(`Failed to create trip at index ${i}:`, error);
    }
  }

  result.success = result.failed === 0;
  return result;
};

export default {
  getTrip,
  getAllTrips,
  getTripsByStatus,
  getTripsByLoadOwner,
  getTripsByLender,
  createTrip,
  updateTrip,
  addBid,
  uploadDocument,
  getTripBids,
  deleteTrip,
  bulkCreateTrips,
};

File: fastapi-backend/app/routers/trips.py
from fastapi import APIRouter, HTTPException, status, Depends, Query
from typing import Optional
from psycopg2.extras import RealDictCursor
import time
import random
import string

from ..models import TripResponse, TripCreate, TripUpdate, BidCreate, DocumentUpload, TripBid
from ..auth import get_current_user
from ..database import get_db_cursor

router = APIRouter(prefix="/trips", tags=["trips"])


def _convert_trip_to_float(trip):
    """Convert trip numeric fields from Decimal to float"""
    result = {
        **trip,
        "distance": float(trip["distance"]),
        "weight": float(trip["weight"]),
        "amount": float(trip["amount"]),
    }
    if trip.get("interest_rate") is not None:
        result["interest_rate"] = float(trip["interest_rate"])
    if trip.get("maturity_days") is not None:
        result["maturity_days"] = int(trip["maturity_days"])
    if trip.get("load_owner_rating") is not None:
        result["load_owner_rating"] = float(trip["load_owner_rating"])
    return result


def _convert_bid_to_float(bid):
    """Convert bid numeric fields from Decimal to float"""
    return {
        **bid,
        "amount": float(bid["amount"]),
        "interest_rate": float(bid["interest_rate"]),
    }


@router.get("/", response_model=list[TripResponse])
async def get_trips(status_filter: Optional[str] = Query(None, alias="status")):
    """Get all trips or filter by status"""
    with get_db_cursor() as cursor:
        if status_filter:
            cursor.execute("SELECT * FROM trips WHERE status = %s ORDER BY created_at DESC", (status_filter,))
        else:
            cursor.execute("SELECT * FROM trips ORDER BY created_at DESC")

        trips = cursor.fetchall()

        # Get all bids for all trips
        cursor.execute("SELECT * FROM trip_bids ORDER BY created_at DESC")
        all_bids = cursor.fetchall()

        # Group bids by trip_id
        bids_by_trip = {}
        for bid in all_bids:
            trip_id = bid["trip_id"]
            if trip_id not in bids_by_trip:
                bids_by_trip[trip_id] = []
            bids_by_trip[trip_id].append(_convert_bid_to_float(dict(bid)))

        # Add bids to trips
        result = []
        for trip in trips:
            trip_dict = _convert_trip_to_float(dict(trip))
            trip_dict["bids"] = bids_by_trip.get(trip["id"])
            result.append(trip_dict)

    return result


@router.get("/{trip_id}", response_model=TripResponse)
async def get_trip(trip_id: str):
    """Get trip by ID"""
    with get_db_cursor() as cursor:
        cursor.execute("SELECT * FROM trips WHERE id = %s", (trip_id,))
        trip = cursor.fetchone()

        if not trip:
            raise HTTPException(status_code=404, detail="Trip not found")

        # Get bids
        cursor.execute("SELECT * FROM trip_bids WHERE trip_id = %s ORDER BY created_at DESC", (trip_id,))
        bids = cursor.fetchall()

        # Get documents
        cursor.execute("SELECT * FROM trip_documents WHERE trip_id = %s", (trip_id,))
        docs = cursor.fetchall()
        documents = {doc["document_type"]: doc["document_data"] for doc in docs}

        trip_dict = _convert_trip_to_float(dict(trip))
        trip_dict["bids"] = [_convert_bid_to_float(dict(bid)) for bid in bids] if bids else None
        trip_dict["documents"] = documents if documents else None

    return trip_dict


@router.post("/", response_model=TripResponse, status_code=status.HTTP_201_CREATED)
async def create_trip(trip: TripCreate):
    """Create new trip"""
    with get_db_cursor(commit=True) as cursor:
        trip_id = f"trip-{int(time.time() * 1000)}-{''.join(random.choices(string.ascii_lowercase + string.digits, k=9))}"

        cursor.execute(
            """
            INSERT INTO trips (
                id, load_owner_id, load_owner_name, load_owner_logo, load_owner_rating,
                client_company, client_logo, origin, destination, distance, load_type,
                weight, amount, interest_rate, maturity_days, risk_level, insurance_status, status
            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, 'pending')
            RETURNING *
            """,
            (
                trip_id, trip.load_owner_id, trip.load_owner_name, trip.load_owner_logo,
                trip.load_owner_rating, trip.client_company, trip.client_logo, trip.origin,
                trip.destination, trip.distance, trip.load_type, trip.weight, trip.amount,
                trip.interest_rate, trip.maturity_days or 30, trip.risk_level or 'low',
                trip.insurance_status
            )
        )

        new_trip = cursor.fetchone()

    return _convert_trip_to_float(dict(new_trip))


@router.put("/{trip_id}", response_model=TripResponse)
async def update_trip(trip_id: str, trip_update: TripUpdate):
    """Update trip"""
    with get_db_cursor(commit=True) as cursor:
        # Check if trip exists
        cursor.execute("SELECT id FROM trips WHERE id = %s", (trip_id,))
        if not cursor.fetchone():
            raise HTTPException(status_code=404, detail="Trip not found")

        # Build update query
        fields = []
        values = []
        for field, value in trip_update.model_dump(exclude_unset=True).items():
            if value is not None:
                fields.append(f"{field} = %s")
                values.append(value)

        if not fields:
            # No fields to update, return current trip
            cursor.execute("SELECT * FROM trips WHERE id = %s", (trip_id,))
            trip = cursor.fetchone()
            return _convert_trip_to_float(dict(trip))

        values.append(trip_id)
        query = f"UPDATE trips SET {', '.join(fields)} WHERE id = %s RETURNING *"

        cursor.execute(query, values)
        updated_trip = cursor.fetchone()

    return _convert_trip_to_float(dict(updated_trip))


@router.post("/{trip_id}/bids", response_model=TripBid, status_code=status.HTTP_201_CREATED)
async def add_bid(trip_id: str, bid: BidCreate, current_user: dict = Depends(get_current_user)):
    """Add bid to trip"""
    with get_db_cursor(commit=True) as cursor:
        # Check if trip exists
        cursor.execute("SELECT id FROM trips WHERE id = %s", (trip_id,))
        if not cursor.fetchone():
            raise HTTPException(status_code=404, detail="Trip not found")

        bid_id = f"bid-{int(time.time() * 1000)}-{''.join(random.choices(string.ascii_lowercase + string.digits, k=9))}"

        cursor.execute(
            """
            INSERT INTO trip_bids (id, trip_id, lender_id, lender_name, amount, interest_rate)
            VALUES (%s, %s, %s, %s, %s, %s)
            RETURNING *
            """,
            (bid_id, trip_id, current_user["user_id"], current_user["name"], bid.amount, bid.interest_rate)
        )

        new_bid = cursor.fetchone()

    return _convert_bid_to_float(dict(new_bid))


@router.get("/{trip_id}/bids", response_model=list[TripBid])
async def get_trip_bids(trip_id: str):
    """Get bids for trip"""
    with get_db_cursor() as cursor:
        cursor.execute("SELECT * FROM trip_bids WHERE trip_id = %s ORDER BY created_at DESC", (trip_id,))
        bids = cursor.fetchall()

    return [_convert_bid_to_float(dict(bid)) for bid in bids]


@router.post("/{trip_id}/documents")
async def upload_document(trip_id: str, document: DocumentUpload, current_user: dict = Depends(get_current_user)):
    """Upload document for trip"""
    with get_db_cursor(commit=True) as cursor:
        # Check if trip exists
        cursor.execute("SELECT id FROM trips WHERE id = %s", (trip_id,))
        if not cursor.fetchone():
            raise HTTPException(status_code=404, detail="Trip not found")

        doc_id = f"doc-{int(time.time() * 1000)}-{''.join(random.choices(string.ascii_lowercase + string.digits, k=9))}"

        # Delete existing document of same type
        cursor.execute(
            "DELETE FROM trip_documents WHERE trip_id = %s AND document_type = %s",
            (trip_id, document.documentType)
        )

        # Insert new document
        cursor.execute(
            """
            INSERT INTO trip_documents (id, trip_id, document_type, document_data, uploaded_by)
            VALUES (%s, %s, %s, %s, %s)
            RETURNING *
            """,
            (doc_id, trip_id, document.documentType, document.documentData, current_user["user_id"])
        )

        new_doc = cursor.fetchone()

    return dict(new_doc)


@router.delete("/{trip_id}")
async def delete_trip(trip_id: str):
    """Delete trip"""
    with get_db_cursor(commit=True) as cursor:
        cursor.execute("DELETE FROM trips WHERE id = %s RETURNING id", (trip_id,))
        result = cursor.fetchone()

    if not result:
        raise HTTPException(status_code=404, detail="Trip not found")

    return {"message": "Trip deleted successfully"}
